modyn/supervisor/internal/pipeline_executor/pipeline_executor.py

```diff
@@ -1,554 +1,855 @@
-            PipelineStage.STORE_TRAINED_MODEL, MsgType.ID, id_submsg(IdType.TRIGGER, trigger_id)
-        )
-        # We store the trained model for evaluation in any case.
-        self._sw.start("store_trained_model", overwrite=True)
-        model_id = self.grpc.store_trained_model(self.current_training_id)
-        self.trigger.inform_previous_model(model_id)
-        self.pipeline_log["supervisor"]["triggers"][trigger_id]["store_trained_model_time"] = self._sw.stop()
+    @pipeline_stage(PipelineType.NEW_BATCH, PipelineStage.EVALUATE_TRIGGER_POLICIES, track=True)
+    def _evaluate_trigger_policies(
+        self, s: ExecutionState, log: StageLog, batch: list[tuple[int, int, int]]
+    ) -> Generator[int, None, None]:
+        """Evaluate trigger policy and inform selector.
 
-        # Only if the pipeline actually wants to continue the training on it, we set previous model.
-        if self.pipeline_config["training"]["use_previous_model"]:
-            self.previous_model_id = model_id
+        Returns:
+            List of indexes of data points that caused a trigger.
+        """
+        try:
+            # add log data
+            log.info = EvaluateTriggerInfo(batch_size=len(batch))
 
-        self.triggers.append(trigger_id)
-        return model_id
+            # Evaluate trigger policy
+            for trigger_index, t_micros in timed_generator(self.trigger.inform(batch)):
+                log.info.trigger_indexes.append(trigger_index)
+                log.info.trigger_eval_times.append(int(t_micros))
+                yield trigger_index
 
-    def _get_trigger_timespan(
-        self, is_first_triggering_data: bool, triggering_data: list[tuple[int, int, int]]
-    ) -> tuple[int, int]:
-        if is_first_triggering_data:
-            # now it is the first trigger in this batch. Triggering_data can be empty.
-            # when it is indeed empty, then there is remaining data in the last batch
-            # because num_samples_in_trigger is not 0.
-            assert len(triggering_data) > 0 or self.remaining_data_range is not None
+        finally:
+            assert log.info
+            logger.info(f"There were {len(log.info.trigger_indexes)} triggers in this batch.")
 
-            if self.remaining_data_range is not None:
-                first_timestamp = self.remaining_data_range[0]
-                last_timestamp = self.remaining_data_range[1] if len(triggering_data) == 0 else triggering_data[-1][1]
-            else:
-                first_timestamp = triggering_data[0][1]
-                last_timestamp = triggering_data[-1][1]
-        else:
-            assert len(triggering_data) > 0
-            # since num_samples_in_trigger is not 0, we are sure that triggering_data is not empty
-            first_timestamp = triggering_data[0][1]
-            last_timestamp = triggering_data[-1][1]
+    @pipeline_stage(PipelineType.NEW_BATCH, PipelineStage.EXECUTE_TRIGGERS, track=True)
+    def _execute_triggers(
```

Comment:
I am not sure whether I like the wording `execute` here. Maybe `handle` (because we _execute_ training jobs, while handling triggers), although handle is also quite general...

Reply:
maybe `invoke`?

Reply:
Okay no, let's just stick with `handle`